### PR TITLE
Keycloak

### PR DESCRIPTION
**Pull request by n1zea144**
This is a merge of the [keycloak support branch](https://github.com/cBioPortal/cbioportal/pull/2284) which has already been reviewed and approved into the master branch.

# Checks
- [ ] Runs on Heroku.
- [ ] Follows [7 rules of great commit messages](http://chris.beams.io/posts/git-commit/). For most PRs a single commit should suffice, in some cases multiple topical commits can be useful. During review it is ok to see tiny commits (e.g. Fix reviewer comments), but right before the code gets merged to master or rc branch, any such commits should be squashed since they are useless to the other developers. Definitely avoid [merge commits, use rebase instead.](http://nathanleclaire.com/blog/2014/09/14/dont-be-scared-of-git-rebase/)
- [ ] Follows the [Google Style Guide](https://github.com/google/styleguide).
- [ ] Make sure your commit messages end with a Signed-off-by string (this line
  can be automatically added by git if you run the `git-commit` command with
  the `-s` option)
- [ ] If this is a feature, the PR is to rc. If this is a bug fix, the PR is to
  hotfix.

# Any screenshots or GIFs?
If this is a new visual feature please add a before/after screenshot or gif
here with e.g. [GifGrabber](http://www.gifgrabber.com/).

# Notify reviewers
Read our [Pull request merging
policy](../CONTRIBUTING.md#pull-request-merging-policy). If you are part of the
cBioPortal organization, notify the approprate team (remove inappropriate):

@cBioPortal/frontend
@cBioPortal/backend
@cBioPortal/devops

If you are not part of the cBioPortal organization look at who worked on the
file before you. Please use `git blame <filename>` to determine that
and notify them here:

### CODE REVIEW

**Comment by n1zea144** [comment]
@mandawilson Can you approve this PR to get the keycloak support into the master branch?

File: security/security-spring/src/main/java/org/cbioportal/security/spring/authentication/keycloak/SAMLUserDetailsServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+    }
+
+    /**
+     * 
+     * @param attributeValue
+     * @return String type attribute value
+     */
+    private String getAttributeValue(XMLObject attributeValue)
+    {
+        return attributeValue == null ?
```
Comment (zheins) [inline security/security-spring/src/main/java/org/cbioportal/security/spring/authentication/keycloak/SAMLUserDetailsServiceImpl.java]: It's hard to debug statements like this. Can we break it out into a few if-else blocks?

File: security/security-spring/src/main/java/org/cbioportal/security/spring/authentication/keycloak/SAMLUserDetailsServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+                List<XMLObject> attributeValues = cAttribute.getAttributeValues();
+                if (!attributeValues.isEmpty()) {
+                    userRoles.add(new StringBuilder(APP_NAME).append(":").append(
+                        getAttributeValue(attributeValues.get(0))).toString());
+                }    
+            }
+        }
+        
+        try {
+
```
Comment (zheins) [inline security/security-spring/src/main/java/org/cbioportal/security/spring/authentication/keycloak/SAMLUserDetailsServiceImpl.java]: Unnecessary whitespace

File: security/security-spring/src/main/java/org/cbioportal/security/spring/authentication/keycloak/SAMLUserDetailsServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+                    userRoles.add(new StringBuilder(APP_NAME).append(":").append(
+                        getAttributeValue(attributeValues.get(0))).toString());
+                }    
+            }
+        }
+        
+        try {
+            
+            if (userId == null) {
+
```
Comment (zheins) [inline security/security-spring/src/main/java/org/cbioportal/security/spring/authentication/keycloak/SAMLUserDetailsServiceImpl.java]: Whitespace

File: security/security-spring/src/main/java/org/cbioportal/security/spring/authentication/keycloak/SAMLUserDetailsServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+    {
+        PortalUserDetails userDetailsObj = null;
+        String userId = null;
+        List<String> userRoles = new ArrayList<String>();
+
+        // get userid and roles: iterate over attributes searching for "email" and "roles":
+        for (Attribute cAttribute : credential.getAttributes()) {
+            String attrName = cAttribute.getName();
+            log.debug("loadUserBySAML(), parsing attribute -" + attrName);
+
```
Comment (zheins) [inline security/security-spring/src/main/java/org/cbioportal/security/spring/authentication/keycloak/SAMLUserDetailsServiceImpl.java]: Extra newline not needed.

File: security/security-spring/src/main/java/org/cbioportal/security/spring/authentication/keycloak/SAMLUserDetailsServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+                String errorMsg = "loadUserBySAML(), can not instantiate PortalUserDetails from SAML assertion." 
+                    + "Expected 'email' attribute was not found or has no values. ";
+                log.error(errorMsg);
+                throw new Exception(errorMsg);
+            }
+
+            log.debug("loadUserBySAML(), IDP successfully authenticated user, userid: " + userId);
+
+            //add granted authorities:
+            if (userRoles.size() > 0) userDetailsObj = new PortalUserDetails(userId,
```
Comment (zheins) [inline security/security-spring/src/main/java/org/cbioportal/security/spring/authentication/keycloak/SAMLUserDetailsServiceImpl.java]: Google style dictates to always use braces even with a single statement. I think it makes the code more clear, but it's not necessary.

**Comment by pieterlukasse** [comment]
@n1zea144 should this have been to `master`? Maybe you can merge it to `rc` as well?

**Comment by jjgao** [comment]
Not sure why this one went directly into `master`. I'll merge it to `hotfix` and `master`.
